File: xdm-app/src/main/java/xdm/app/ui/screens/AppWindow.java
package xdm.app.ui.screens;

import java.awt.*;
import java.awt.desktop.AppReopenedListener;
import java.awt.event.*;
import java.util.*;
import java.util.List;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.border.MatteBorder;
import javax.swing.event.*;

import com.formdev.flatlaf.util.SystemInfo;
import xdm.app.AppContext;
import xdm.app.constants.AppConstants;
//import xdm.app.models.DownloadEntry;
import xdm.app.data.DbRecord;
import xdm.app.ui.components.*;
import xdman.Config;
// import xdman.XDMApp;
import xdman.ui.res.*;
import xdman.util.*;

public class AppWindow extends JFrame implements ActionListener {
  private MainListView listView;
  private JTextField txtSearch;
  private JPopupMenu popupCtx;
  private JMenu startQMenu, stopQMenu;
  private int windowState;

  public AppWindow(Image image) {
    if (SystemInfo.isWindows) {
      getRootPane()
          .putClientProperty(
              "JRootPane.titleBarBackground", UIManager.getColor("Table.background"));
    }
    setIconImage(image);
    if (SystemInfo.isMacOS) {
      applyMacOSWindowCustomizations(image);
    }
    setTitle(AppConstants.XDM_WINDOW_TITLE);
    setWindowSizeAndPosition();
    initWindow();
  }

  private void initWindow() {
    this.listView = new MainListView();
    //    this.createPopupMenu();
    //    this.listView.installPopupMenu(this.popupCtx);
    var filterPanel = new FilterListPanel();
    var toolbar = new AppToolBar(s -> {}, this);
    toolbar.setMultiSelectView(false);
    listView.setSelectModeCallback(toolbar::setMultiSelectView);
    //    var filterPanel =
    //        new FilterPanel(
    //            f -> {
    //              toolbar.updateButtons(Collections.emptyList());
    //              // listView.setFilterItem(f);
    //            });
    // this.listView.setSelectionCallback(list ->
    // toolbar.updateButtons(listView.getSelectedItems()));

    var panel = new JPanel(new BorderLayout(10, 0));
    panel.add(toolbar.getComponent(), BorderLayout.NORTH);
    panel.add(listView.getComponent());
    panel.setBorder(new EmptyBorder(7, 0, 0, 0));
    var splitPane = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT);
    splitPane.setBorder(new MatteBorder(1, 0, 0, 0, Color.BLACK));
    splitPane.setDividerLocation(180);
    splitPane.setLeftComponent(filterPanel.getComponent());
    splitPane.setRightComponent(panel);
    add(splitPane);

    // createPopupMenu();

    ToolTipManager.sharedInstance().setInitialDelay(500);

    addWindowListener(
        new WindowAdapter() {
          @Override
          public void windowOpened(WindowEvent e) {
            // listView.focus();
            windowState = AppWindow.this.getExtendedState();
          }

          @Override
          public void windowStateChanged(WindowEvent e) {
            if (AppWindow.this.getExtendedState() != JFrame.ICONIFIED) {
              windowState = AppWindow.this.getExtendedState();
            }
          }
        });
  }

  public void updateDownloadInView(int index) {
    this.listView.rowUpdated(index);
  }

  public void addDownloadInView(int index) {
    this.listView.rowAdded(index);
  }

  public void restoreWindowIfNeeded() {
    //    if (this.getExtendedState() == JFrame.ICONIFIED) {
    //      this.setExtendedState(this.windowState);
    //    }
    if (!this.isVisible()) {
      setVisible(true);
    }
  }

  @Override
  public void actionPerformed(ActionEvent e) {
    if (e.getSource() instanceof JComponent c) {
      var name = c.getName();
      if (name == null) {
        return;
      }

      switch (name) {
        case "TOOL_DOWNLOAD":
          AppContext.INSTANCE.getApp().addDownload(null);
          return;
      }

      if (name.startsWith("STOP")) {
        AppMenuHandler.stopQueue(name);
      } else if (name.startsWith("START")) {
        AppMenuHandler.startQueue(name);
      } else if ("TOOL_DOWNLOAD".equals(name) || "MENU_ADD_URL".equals(name)) {
        AppContext.INSTANCE.getApp().addDownload(null);
      } else if ("PAUSE".equals(name) || "MENU_PAUSE".equals(name)) {
        // AppMenuHandler.pauseDownloads(this);
      } else if ("CTX_COPY_URL".equals(name)) {
        AppMenuHandler.copyUrl(this);
      } else if ("LBL_SHOW_PROGRESS".equals(name)) {
        AppMenuHandler.showProgressWindow(this);
      } else if ("MENU_RESTART".equals(name)) {
        //        AppMenuHandler.restartDownloads(this);
      } else if ("RESUME".equals(name) || "MENU_RESUME".equals(name)) {
        //        AppMenuHandler.resumeDownloads(this);
      } else if ("CTX_OPEN_FILE".equals(name)) {
        AppMenuHandler.openFile(this);
      } else if ("CTX_OPEN_FOLDER".equals(name)) {
        //        AppMenuHandler.openFolder(this);
      } else if ("MENU_EXIT".equals(name)) {
        // XDMApp.getInstance().exit();
      } else if ("MENU_OPTIONS".equals(name) || "OPTIONS".equals(name)) {
        // SettingsPage.getInstance().showPanel(this, "PG_SETTINGS");
      } else if ("MENU_REFRESH_LINK".equals(name)) {
        AppMenuHandler.openRefreshPage(this);
      } else if ("MENU_PROPERTIES".equals(name)) {
        AppMenuHandler.showProperties(this);
      } else if ("MENU_BROWSER_INT".equals(name)) {
        // SettingsPage.getInstance().showPanel(this, "BTN_MONITORING");
      } else if ("MENU_SPEED_LIMITER".equals(name)) {
        //        int ret = SpeedLimiter.getSpeedLimit();
        //        if (ret >= 0) {
        //          Config.getInstance().setSpeedLimit(ret);
        //        }
      } else if ("DESC_Q_TITLE".equals(name)) {
        // SettingsPage.getInstance().showPanel(this, "Q_MAN");
      } else if ("MENU_DELETE_DWN".equals(name)
          || "DELETE".equals(name)
          || "DESC_DEL".equals(name)) {
        //        AppMenuHandler.deleteDownloads(this);
      } else if ("MENU_DELETE_COMPLETED".equals(name)) {
        AppMenuHandler.deleteCompleted(this);
      } else if ("MENU_ABOUT".equals(name)) {
        //				AboutPage aboutPage = new AboutPage(this);
        //				aboutPage.showPanel();
      } else if ("CTX_SAVE_AS".equals(name)) {
        AppMenuHandler.changeFile(this);
      } else if ("MENU_IMPORT".equals(name)) {
        //        JFileChooser jfc = new JFileChooser();
        //        if (jfc.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
        //          File file = jfc.getSelectedFile();
        //          XDMApp.getInstance().loadDownloadList(file);
        //        }
      } else if ("MENU_EXPORT".equals(name)) {
        //        JFileChooser jfc = new JFileChooser();
        //        if (jfc.showSaveDialog(this) == JFileChooser.APPROVE_OPTION) {
        //          File file = jfc.getSelectedFile();
        //          XDMApp.getInstance().saveDownloadList(file);
        //        }
      } else if ("MENU_CONTENTS".equals(name)) {
        //        XDMUtils.browseURL(XDMApp.APP_WIKI_URL);
      } else if ("MENU_HOME_PAGE".equals(name)) {
        //        XDMUtils.browseURL(XDMApp.APP_HOME_URL);
      } else if ("MENU_UPDATE".equals(name)) {
        //        XDMUtils.browseURL(XDMApp.APP_UPDATE_CHK_URL + XDMApp.APP_VERSION);
      } else if ("MENU_LANG".equals(name)) {
        AppMenuHandler.showLanguageDlg(this);
      } else if ("MENU_BATCH_DOWNLOAD".equals(name)) {
        AppMenuHandler.showBatchPatternDialog();
      } else if ("MENU_CLIP_ADD_MENU".equals(name)) {
        AppMenuHandler.showBatchDialog(this);
      } else if ("LBL_OPTIMIZE_NETWORK".equals(name)) {
        AppMenuHandler.optimizeRWin();
      } else if ("LBL_TRANSLATE".equals(name)) {
        AppMenuHandler.openTranslationPage();
      } else if ("LBL_SUPPORT_PAGE".equals(name)) {
        AppMenuHandler.openSupportPage();
      } else if ("LBL_REPORT_PROBLEM".equals(name)) {
        AppMenuHandler.openBugReportPage();
      }
    }
  }

  private void createMainMenu() {

    JMenuBar bar = new JMenuBar();

    JMenu file = new JMenu(StringResource.get("MENU_FILE"));

    addMenuItem("MENU_ADD_URL", file);
    addMenuItem("MENU_VIDEO_DWN", file);
    addMenuItem("MENU_CLIP_ADD_MENU", file);
    addMenuItem("MENU_BATCH_DOWNLOAD", file);
    addMenuItem("MENU_DELETE_DWN", file);
    addMenuItem("MENU_DELETE_COMPLETED", file);
    addMenuItem("MENU_EXPORT", file);
    addMenuItem("MENU_IMPORT", file);
    addMenuItem("MENU_EXIT", file);

    JMenu dwn = new JMenu(StringResource.get("MENU_DOWNLOAD"));

    addMenuItem("MENU_PAUSE", dwn);
    addMenuItem("MENU_RESUME", dwn);
    addMenuItem("MENU_RESTART", dwn);
    addMenuItem("DESC_Q_TITLE", dwn);

    PopupMenuListener popupListener =
        new PopupMenuListener() {

          @Override
          public void popupMenuWillBecomeVisible(PopupMenuEvent e) {
            loadQueueMenu(startQMenu);
            loadQueueMenu(stopQMenu);
          }

          @Override
          public void popupMenuWillBecomeInvisible(PopupMenuEvent e) {}

          @Override
          public void popupMenuCanceled(PopupMenuEvent e) {}
        };

    startQMenu = addSubMenu("MENU_START_Q", dwn, popupListener);
    stopQMenu = addSubMenu("MENU_STOP_Q", dwn, popupListener);

    JMenu tools = new JMenu(StringResource.get("MENU_TOOLS"));

    addMenuItem("MENU_OPTIONS", tools);
    addMenuItem("MENU_REFRESH_LINK", tools);
    addMenuItem("MENU_PROPERTIES", tools);
    addMenuItem("MENU_SPEED_LIMITER", tools);
    addMenuItem("MENU_LANG", tools);
    addMenuItem("MENU_MEDIA_CONVERTER", tools);
    addMenuItem("LBL_OPTIMIZE_NETWORK", tools);
    addMenuItem("MENU_BROWSER_INT", tools);

    JMenu help = new JMenu(StringResource.get("MENU_HELP"));
    addMenuItem("MENU_CONTENTS", help);
    addMenuItem("MENU_HOME_PAGE", help);
    addMenuItem("LBL_SUPPORT_PAGE", help);
    addMenuItem("LBL_REPORT_PROBLEM", help);
    addMenuItem("LBL_TRANSLATE", help);
    addMenuItem("MENU_UPDATE", help);
    addMenuItem("MENU_ABOUT", help);

    bar.add(file);
    bar.add(dwn);
    bar.add(tools);
    bar.add(help);

    setJMenuBar(bar);
  }

  private void addMenuItem(String id, JComponent menu) {
    var mItem = new JMenuItem(StringResource.get(id));
    mItem.setName(id);
    mItem.addActionListener(this);
    menu.add(mItem);
  }

  private JMenu addSubMenu(String id, JMenu parentMenu, PopupMenuListener popupListener) {
    JMenu menu = new JMenu(StringResource.get(id));
    menu.setName(id);
    menu.addActionListener(this);
    menu.getPopupMenu().addPopupMenuListener(popupListener);
    parentMenu.add(menu);
    return menu;
  }

  private void setWindowSizeAndPosition() {
    if (Config.getInstance().getWidth() < 0 || Config.getInstance().getHeight() < 0)
      setSize(800, 500);
    if (Config.getInstance().getX() < 0 || Config.getInstance().getY() < 0)
      setLocationRelativeTo(null);
  }

  private void loadQueueMenu(JMenu menu) {
    if (menu.getName().equals("MENU_START_Q")) {
      loadStartQueueMenu(menu);
    } else if (menu.getName().equals("MENU_STOP_Q")) {
      loadStopQueueMenu(menu);
    }
  }

  private void loadStopQueueMenu(JMenu menu) {
    //    menu.removeAll();
    //    ArrayList<DownloadQueue> queues = XDMApp.getInstance().getQueueList();
    //    for (int i = 0; i < queues.size(); i++) {
    //      DownloadQueue q = queues.get(i);
    //      if (q.isRunning()) {
    //        JMenuItem mitem = new JMenuItem(q.getName());
    //        mitem.setForeground(ColorResource.getLightFontColor());
    //        mitem.setName("STOP:" + q.getQueueId());
    //        mitem.addActionListener(this);
    //        menu.add(mitem);
    //      }
    //    }
  }

  private void loadStartQueueMenu(JMenu menu) {
    //    menu.removeAll();
    //    ArrayList<DownloadQueue> queues = XDMApp.getInstance().getQueueList();
    //    for (int i = 0; i < queues.size(); i++) {
    //      DownloadQueue q = queues.get(i);
    //      if (!q.isRunning()) {
    //        JMenuItem mitem = new JMenuItem(q.getName());
    //        mitem.setForeground(ColorResource.getLightFontColor());
    //        mitem.setName("START:" + q.getQueueId());
    //        mitem.addActionListener(this);
    //        menu.add(mitem);
    //      }
    //    }
  }

  private void createPopupMenu() {
    popupCtx = new JPopupMenu();
    addMenuItem("CTX_OPEN_FILE", popupCtx);
    addMenuItem("CTX_OPEN_FOLDER", popupCtx);
    addMenuItem("CTX_SAVE_AS", popupCtx);
    addMenuItem("MENU_PAUSE", popupCtx);
    addMenuItem("MENU_RESUME", popupCtx);
    addMenuItem("MENU_DELETE_DWN", popupCtx);
    addMenuItem("MENU_REFRESH_LINK", popupCtx);
    addMenuItem("LBL_SHOW_PROGRESS", popupCtx);
    addMenuItem("CTX_COPY_URL", popupCtx);
    addMenuItem("CTX_COPY_FILE", popupCtx);
    addMenuItem("MENU_PROPERTIES", popupCtx);
    //    popupCtx.setInvoker(listView.getComponent());
    //    listView.installPopupMenu(popupCtx, this);
  }

  public List<DbRecord> getSelectedDownloads() {
    return Collections.emptyList(); // listView.getSelectedItems();
  }

  private void applyMacOSWindowCustomizations(Image image) {
    /* Set Dock icon in macOS */
    try {
      Taskbar.getTaskbar().setIconImage(image);
    } catch (final UnsupportedOperationException | SecurityException e) {
      // Nothing to do
    }
    try {
      if (SystemInfo.isMacFullWindowContentSupported) {
        this.getRootPane().putClientProperty("apple.awt.transparentTitleBar", true);
      }
    } catch (Exception ex) {
      // Nothing to do
    }
    try {
      Desktop.getDesktop()
          .addAppEventListener(
              (AppReopenedListener) e -> AppContext.INSTANCE.getApp().showAppWindow());
    } catch (Exception ex) {
      // Nothing to do
    }
  }
}

File: xdm-app/src/main/java/xdm/app/ui/screens/AppWindow.kt
package xdm.app.ui.screens

import com.formdev.flatlaf.FlatLaf
import xdm.app.AppContext
import xdm.app.AppContext.app
import xdm.app.XDM_WINDOW_TITLE
import xdm.app.DbRecord
import xdm.app.I8N.text
import xdm.app.OS
import xdm.app.ui.components.AppMenuHandler
import xdm.app.ui.components.AppToolBar
import xdm.app.ui.components.FilterListPanel
import xdm.app.ui.components.MainListView
import xdm.app.ui.components.UpdatePanel
import xdm.app.update.UpdateChecker
import xdm.app.utils.applyMacOSWindowCustomizations
import xdm.app.utils.detectOS
import xdm.core.util.Logger
import java.awt.*
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.*
import javax.swing.border.EmptyBorder
import javax.swing.border.MatteBorder
import javax.swing.event.PopupMenuEvent
import javax.swing.event.PopupMenuListener

class AppWindow(image: Image) : JFrame(), ActionListener {
    private val listView = MainListView()
    private val updatePanel = UpdatePanel()

    init {
        title = XDM_WINDOW_TITLE
        iconImage = image
        val os = detectOS()
        if (os == OS.Windows) {
            getRootPane().putClientProperty("JRootPane.titleBarBackground", UIManager.getColor("Table.background"))
        } else if (os == OS.MacOS) {
            applyMacOSWindowCustomizations(image)
        }

        setWindowSizeAndPosition()
        initWindow()
        checkForUpdates()

        //ProgressWindow().isVisible = true
    }

    /**
     * Kicks off a background update check at startup; if a newer release is
     * found, reveals the update banner at the bottom of the window (on the EDT).
     */
    private fun checkForUpdates() {
        UpdateChecker.checkForUpdate { info ->
            if (info != null) {
                SwingUtilities.invokeLater {
                    updatePanel.showUpdate(info.latestVersion, info.downloadUrl)
                }
            }
        }
    }

    private fun initWindow() {
        val filterPanel = FilterListPanel(
            stateChanged = { listView.filterStateChanged(it) },
            categoryChanged = { listView.filterCategoryChanged(it) }
        )
        val toolbar = AppToolBar(
            { listView.searchTextChanged(it) },
            this,
            { key, asc -> listView.sort(key, asc) },
            { filterPanel.reloadCategories() }
        )
        toolbar.setMultiSelectView(false)
        listView.selectModeCallback = { toolbar.setMultiSelectView(it) }

        val panel = JPanel(BorderLayout(10, 0)).apply {
            add(toolbar.component, BorderLayout.NORTH)
            add(listView.component)
            border = EmptyBorder(7, 0, 0, 0)
            if (AppContext.config.theme == "light") {
                background = UIManager.getColor("Table.background")
            }
        }

        // Keep the black top border in the dark theme (as before); use FlatLaf's
        // border color in the light theme so it isn't a harsh black line.
        val topBorderColor = if (FlatLaf.isLafDark()) {
            Color.BLACK
        } else {
            UIManager.getColor("Component.borderColor") ?: Color.GRAY
        }
        val splitPane = JSplitPane(JSplitPane.HORIZONTAL_SPLIT).apply {
            border = MatteBorder(1, 0, 0, 0, topBorderColor)
            dividerLocation = 180
            leftComponent = filterPanel.component
            rightComponent = panel
            if (AppContext.config.theme == "light") {
                background = UIManager.getColor("Table.background")
            }
        }
        add(splitPane, BorderLayout.CENTER)
        add(updatePanel, BorderLayout.SOUTH)

        ToolTipManager.sharedInstance().initialDelay = 500
    }

    fun updateDownloadInView(index: Int) {
        listView.rowUpdated(index)
    }

    fun deleteDownloadInView(index: Int) {
        listView.rowDeleted(index)
    }

    fun addDownloadInView(index: Int) {
        listView.rowAdded(index)
    }


    override fun actionPerformed(e: ActionEvent) {
        Logger.info("XDM", "Command: ${e.actionCommand}")
        if (e.source is JComponent) {
            val name: String = (e.source as JComponent).name ?: return

            when (name) {
                "TOOL_DOWNLOAD" -> {
                    app.addDownload(null)
                    return
                }

                "TOOL_CLEAR" -> {
                    clearDownloads()
                }
            }

            if (name.startsWith("STOP")) {
                AppMenuHandler.stopQueue(name)
            } else if (name.startsWith("START")) {
                AppMenuHandler.startQueue(name)
            } else if ("TOOL_DOWNLOAD" == name || "MENU_ADD_URL" == name) {
                app.addDownload(null)
            } else if ("PAUSE" == name || "MENU_PAUSE" == name) {
                // AppMenuHandler.pauseDownloads(this);
            } else if ("LBL_SHOW_PROGRESS" == name) {
                AppMenuHandler.showProgressWindow(this)
            } else if ("MENU_RESTART" == name) {
                //        AppMenuHandler.restartDownloads(this);
            } else if ("RESUME" == name || "MENU_RESUME" == name) {
                //        AppMenuHandler.resumeDownloads(this);
            } else if ("CTX_OPEN_FILE" == name) {
                AppMenuHandler.openFile(this)
            } else if ("CTX_OPEN_FOLDER" == name) {
                //        AppMenuHandler.openFolder(this);
            } else if ("MENU_EXIT" == name) {
                // XDMApp.getInstance().exit();
            } else if ("MENU_OPTIONS" == name || "OPTIONS" == name) {
                // SettingsPage.getInstance().showPanel(this, "PG_SETTINGS");
            } else if ("MENU_REFRESH_LINK" == name) {
                AppMenuHandler.openRefreshPage(this)
            } else if ("MENU_PROPERTIES" == name) {
                //AppMenuHandler.showProperties(this)
            } else if ("MENU_BROWSER_INT" == name) {
                // SettingsPage.getInstance().showPanel(this, "BTN_MONITORING");
            } else if ("MENU_SPEED_LIMITER" == name) {
                //        int ret = SpeedLimiter.getSpeedLimit();
                //        if (ret >= 0) {
                //          Config.getInstance().setSpeedLimit(ret);
                //        }
            } else if ("DESC_Q_TITLE" == name) {
                // SettingsPage.getInstance().showPanel(this, "Q_MAN");
            } else if ("TOOL_DELETE" == name) {
                Logger.info("Selected items: ${listView.selectedItems}")
                AppMenuHandler.deleteSelectedDownloads(listView.selectedItems, this)
            } else if ("MENU_DELETE_COMPLETED" == name) {
                AppMenuHandler.deleteCompleted(this)
            } else if ("MENU_ABOUT" == name) {
                //				AboutPage aboutPage = new AboutPage(this);
                //				aboutPage.showPanel();
            } else if ("CTX_SAVE_AS" == name) {
                AppMenuHandler.changeFile(this)
            } else if ("MENU_IMPORT" == name) {
                //        JFileChooser jfc = new JFileChooser();
                //        if (jfc.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
                //          File file = jfc.getSelectedFile();
                //          XDMApp.getInstance().loadDownloadList(file);
                //        }
            } else if ("MENU_EXPORT" == name) {
                //        JFileChooser jfc = new JFileChooser();
                //        if (jfc.showSaveDialog(this) == JFileChooser.APPROVE_OPTION) {
                //          File file = jfc.getSelectedFile();
                //          XDMApp.getInstance().saveDownloadList(file);
                //        }
            } else if ("MENU_CONTENTS" == name) {
                //        XDMUtils.browseURL(XDMApp.APP_WIKI_URL);
            } else if ("MENU_HOME_PAGE" == name) {
                //        XDMUtils.browseURL(XDMApp.APP_HOME_URL);
            } else if ("MENU_UPDATE" == name) {
                //        XDMUtils.browseURL(XDMApp.APP_UPDATE_CHK_URL + XDMApp.APP_VERSION);
            } else if ("MENU_LANG" == name) {
                AppMenuHandler.showLanguageDlg(this)
            } else if ("MENU_BATCH_DOWNLOAD" == name) {
                AppMenuHandler.showBatchPatternDialog()
            } else if ("MENU_CLIP_ADD_MENU" == name) {
                AppMenuHandler.showBatchDialog(this)
            } else if ("LBL_OPTIMIZE_NETWORK" == name) {
                AppMenuHandler.optimizeRWin()
            } else if ("LBL_TRANSLATE" == name) {
                AppMenuHandler.openTranslationPage()
            } else if ("LBL_SUPPORT_PAGE" == name) {
                AppMenuHandler.openSupportPage()
            } else if ("LBL_REPORT_PROBLEM" == name) {
                AppMenuHandler.openBugReportPage()
            }
        }
    }

    private fun clearDownloads() {
        listView.clear()
    }

    private fun createMainMenu() {
        val bar = JMenuBar()

        val file = JMenu(text("MENU_FILE"))

        addMenuItem("MENU_ADD_URL", file)
        addMenuItem("MENU_VIDEO_DWN", file)
        addMenuItem("MENU_CLIP_ADD_MENU", file)
        addMenuItem("MENU_BATCH_DOWNLOAD", file)
        addMenuItem("MENU_DELETE_DWN", file)
        addMenuItem("MENU_DELETE_COMPLETED", file)
        addMenuItem("MENU_EXPORT", file)
        addMenuItem("MENU_IMPORT", file)
        addMenuItem("MENU_EXIT", file)

        val dwn = JMenu(text("MENU_DOWNLOAD"))

        addMenuItem("MENU_PAUSE", dwn)
        addMenuItem("MENU_RESUME", dwn)
        addMenuItem("MENU_RESTART", dwn)
        addMenuItem("DESC_Q_TITLE", dwn)

        val popupListener: PopupMenuListener =
            object : PopupMenuListener {
                override fun popupMenuWillBecomeVisible(e: PopupMenuEvent) {
//                    loadQueueMenu(startQMenu!!)
//                    loadQueueMenu(stopQMenu!!)
                }

                override fun popupMenuWillBecomeInvisible(e: PopupMenuEvent) {}

                override fun popupMenuCanceled(e: PopupMenuEvent) {}
            }

//        startQMenu = addSubMenu("MENU_START_Q", dwn, popupListener)
//        stopQMenu = addSubMenu("MENU_STOP_Q", dwn, popupListener)

        val tools = JMenu(text("MENU_TOOLS"))

        addMenuItem("MENU_OPTIONS", tools)
        addMenuItem("MENU_REFRESH_LINK", tools)
        addMenuItem("MENU_PROPERTIES", tools)
        addMenuItem("MENU_SPEED_LIMITER", tools)
        addMenuItem("MENU_LANG", tools)
        addMenuItem("MENU_MEDIA_CONVERTER", tools)
        addMenuItem("LBL_OPTIMIZE_NETWORK", tools)
        addMenuItem("MENU_BROWSER_INT", tools)

        val help = JMenu(text("MENU_HELP"))
        addMenuItem("MENU_CONTENTS", help)
        addMenuItem("MENU_HOME_PAGE", help)
        addMenuItem("LBL_SUPPORT_PAGE", help)
        addMenuItem("LBL_REPORT_PROBLEM", help)
        addMenuItem("LBL_TRANSLATE", help)
        addMenuItem("MENU_UPDATE", help)
        addMenuItem("MENU_ABOUT", help)

        bar.add(file)
        bar.add(dwn)
        bar.add(tools)
        bar.add(help)

        jMenuBar = bar
    }

    private fun addMenuItem(id: String, menu: JComponent) {
        val mItem = JMenuItem(text(id))
        mItem.name = id
        mItem.addActionListener(this)
        menu.add(mItem)
    }

    private fun addSubMenu(id: String, parentMenu: JMenu, popupListener: PopupMenuListener): JMenu {
        val menu = JMenu(text(id))
        menu.name = id
        menu.addActionListener(this)
        menu.popupMenu.addPopupMenuListener(popupListener)
        parentMenu.add(menu)
        return menu
    }

    private fun setWindowSizeAndPosition() {
        setSize(800, 500)
        setLocationRelativeTo(null)
//        if (Config.getInstance().width < 0 || Config.getInstance().height < 0) setSize(800, 500)
//        if (Config.getInstance().x < 0 || Config.getInstance().y < 0) setLocationRelativeTo(null)
    }

    private fun loadQueueMenu(menu: JMenu) {
        if (menu.name == "MENU_START_Q") {
            loadStartQueueMenu(menu)
        } else if (menu.name == "MENU_STOP_Q") {
            loadStopQueueMenu(menu)
        }
    }

    private fun loadStopQueueMenu(menu: JMenu) {
        //    menu.removeAll();
        //    ArrayList<DownloadQueue> queues = XDMApp.getInstance().getQueueList();
        //    for (int i = 0; i < queues.size(); i++) {
        //      DownloadQueue q = queues.get(i);
        //      if (q.isRunning()) {
        //        JMenuItem mitem = new JMenuItem(q.getName());
        //        mitem.setForeground(ColorResource.getLightFontColor());
        //        mitem.setName("STOP:" + q.getQueueId());
        //        mitem.addActionListener(this);
        //        menu.add(mitem);
        //      }
        //    }
    }

    private fun loadStartQueueMenu(menu: JMenu) {
        //    menu.removeAll();
        //    ArrayList<DownloadQueue> queues = XDMApp.getInstance().getQueueList();
        //    for (int i = 0; i < queues.size(); i++) {
        //      DownloadQueue q = queues.get(i);
        //      if (!q.isRunning()) {
        //        JMenuItem mitem = new JMenuItem(q.getName());
        //        mitem.setForeground(ColorResource.getLightFontColor());
        //        mitem.setName("START:" + q.getQueueId());
        //        mitem.addActionListener(this);
        //        menu.add(mitem);
        //      }
        //    }
    }

    private fun createPopupMenu() {
//        popupCtx = JPopupMenu()
//        addMenuItem("CTX_OPEN_FILE", popupCtx!!)
//        addMenuItem("CTX_OPEN_FOLDER", popupCtx!!)
//        addMenuItem("CTX_SAVE_AS", popupCtx!!)
//        addMenuItem("MENU_PAUSE", popupCtx!!)
//        addMenuItem("MENU_RESUME", popupCtx!!)
//        addMenuItem("MENU_DELETE_DWN", popupCtx!!)
//        addMenuItem("MENU_REFRESH_LINK", popupCtx!!)
//        addMenuItem("LBL_SHOW_PROGRESS", popupCtx!!)
//        addMenuItem("CTX_COPY_URL", popupCtx!!)
//        addMenuItem("CTX_COPY_FILE", popupCtx!!)
//        addMenuItem("MENU_PROPERTIES", popupCtx!!)
        //    popupCtx.setInvoker(listView.getComponent());
        //    listView.installPopupMenu(popupCtx, this);
    }

    val selectedDownloads: List<DbRecord>
        get() = emptyList() // listView.getSelectedItems();

}
